### PR TITLE
Rewrote Machine Processing API

### PR DESCRIPTION
**Pull request by TheBusyBiscuit**
## Description

This PR proposes a new Machine Processing API.
It may also benefit performance a little since it reduces everything into a single map.
The map is also thread-safe thanks to the ReadWriteLock used.

## Changes

* Removed all static "processing" fields
* Removed all static "progress" fields
* Added proper Processing API

## Checklist

- [ ] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

**Comment by TheBusyBiscuit** [comment]
> Thank you for contributing to this project! ❤️

You're welcome, bot ❤️

**Comment by TheBusyBiscuit** [comment]
This is mostly finished.
It still needs to be tested though and perhaps some Unit Tests could be made too.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/geo/GEOMiner.java [hunk 1]
```diff
@@ -42,22 +48,33 @@
...
  */
-public class GEOMiner extends AContainer implements RecipeDisplayItem, HologramOwner {
+public class GEOMiner extends SlimefunItem implements RecipeDisplayItem, EnergyNetComponent, InventoryBlock, HologramOwner, MachineProcessHolder<MiningOperation> {
 
     private static final int[] BORDER = { 0, 1, 2, 3, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 26, 27, 35, 36, 44, 45, 53 };
     private static final int[] OUTPUT_BORDER = { 19, 20, 21, 22, 23, 24, 25, 28, 34, 37, 43, 46, 47, 48, 49, 50, 51, 52 };
     private static final int[] OUTPUT_SLOTS = { 29, 30, 31, 32, 33, 38, 39, 40, 41, 42 };
+
     private static final int PROCESSING_TIME = 14;
+    private static final int ENERGY_CONSUMPTION = 24;
```
Comment (Mooy1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/geo/GEOMiner.java:58]: Having this as a static field breaks anything that wants to extend GeoMiner and change consumption, it should be provided in constructor or smth

**Review by Mooy1 — CHANGES_REQUESTED** [review]

**Comment by Mooy1** [comment]
Also, what is the MachineProcessorHolder meant to do?

**Comment by TheBusyBiscuit** [comment]
@Slimefun/code-reviewers This has been open since January and in need of a review :eyes:

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/machines/MachineProcessor.java [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+     * 
+     * @param pos
+     *            The {@link BlockPosition} of our machine
+     * @param operation
+     *            The {@link MachineOperation} to start
+     * 
+     * @return Whether the {@link MachineOperation} was successfully started. This will return false if another
+     *         {@link MachineOperation} has already been started at that {@link BlockPosition}.
+     */
+    private boolean startOperation(@Nonnull BlockPosition pos, @Nonnull T operation) {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/machines/MachineProcessor.java:137]: pls make this public, I use BlockPosition in addons.

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/machines/MachineProcessor.java [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+     * We don't need to validate our input here as that is already
+     * covered in our public methods.
+     * 
+     * @param pos
+     *            The {@link BlockPosition} at which our machine is located.
+     * 
+     * @return The current {@link MachineOperation} or null.
+     */
+    @Nullable
+    private T getOperation(@Nonnull BlockPosition pos) {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/machines/MachineProcessor.java:188]: pls make this public, I use BlockPosition in addons.

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/machines/MachineProcessor.java [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+     * We don't need to validate our input here as that is already
+     * covered in our public methods.
+     * 
+     * @param pos
+     *            The {@link BlockPosition} at which our machine is located.
+     * 
+     * @return Whether the {@link MachineOperation} was successfully ended. This will return false if there was no
+     *         {@link MachineOperation} to begin with.
+     */
+    private boolean endOperation(@Nonnull BlockPosition pos) {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/machines/MachineProcessor.java:239]: pls make this public, I use BlockPosition in addons.

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/machines/MachineProcessor.java [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+                    Event event = new AsyncMachineOperationFinishEvent(pos, this, operation);
+                    Bukkit.getPluginManager().callEvent(event);
+                }
+
+                return true;
+            } else {
+                return false;
+            }
+        } finally {
+            lock.writeLock().unlock();
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/machines/MachineProcessor.java:260]: We're keeping this locked way longer than needed. This should be unlocked as soon as remove is done. Other plugin processing is very unpredictable

**Review by WalshyDev — APPROVED** [review]
This is looking really good, will be a nice memory reduction too

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/machines/MachineProcessor.java (hunk 5 above)
Comment (TheBusyBiscuit) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/machines/MachineProcessor.java:260]: good point

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/geo/GEOMiner.java [hunk 6]
```diff
@@ -161,27 +172,40 @@ public boolean onClick(InventoryClickEvent e, Player p, int slot, ItemStack curs
     }
 
     @Override
+    public void preRegister() {
+        addItemHandler(new BlockTicker() {
+
+            @Override
+            public void tick(Block b, SlimefunItem sf, Config data) {
+                GEOMiner.this.tick(b);
```
Comment (svr333) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/geo/GEOMiner.java:180]: this syntax is weirding me out

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/operations/FuelOperation.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        Validate.notNull(ingredient, "The Ingredient cannot be null");
+        Validate.isTrue(totalTicks > 0, "The amount of total ticks must be a positive integer");
+
+        this.ingredient = ingredient;
+        this.result = result;
+        this.totalTicks = totalTicks;
+    }
+
+    @Override
+    public void addProgress(int num) {
```
Comment (svr333) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/operations/FuelOperation.java:34]: Maybe a check to make sure its a positive number?
Or start using a uint I suppose

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/operations/CraftingOperation.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        Validate.isTrue(totalTicks > 0, "The amount of total ticks must be a positive integer");
+
+        this.ingredients = ingredients;
+        this.results = results;
+        this.totalTicks = totalTicks;
+    }
+
+    @Override
+    public void addProgress(int num) {
+        currentTicks += num;
```
Comment (svr333) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/operations/CraftingOperation.java:35]: Maybe a check to make sure its a positive number?
Or start using a uint I suppose

Unless you want to support removing progress from machines, that is

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/operations/MiningOperation.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    public MiningOperation(@Nonnull ItemStack result, int totalTicks) {
+        Validate.notNull(result, "The result cannot be null");
+        Validate.isTrue(totalTicks > 0, "The amount of total ticks must be a positive integer");
+
+        this.result = result;
+        this.totalTicks = totalTicks;
+    }
+
+    @Override
+    public void addProgress(int num) {
```
Comment (svr333) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/operations/MiningOperation.java:26]: Maybe a check to make sure its a positive number?
Or start using a uint I suppose

**Review by svr333 — APPROVED** [review]
Code looks good, didnt find any logic errors

**Review by Sfiguz7 — DISMISSED** [review]
Luckily I had already reviewed a good chunk of this lol LGTM

**Review by svr333 — DISMISSED** [review]

**Comment by svr333** [comment]
I think I fucked up with the merge conflict thingy

**Review by svr333 — APPROVED** [review]

**Review by Sfiguz7 — APPROVED** [review]

**Pull request merged**